Support Custom Sentiment Analysis task (LRO) in AnalyzeText Language API

jessicawrr opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

jessicawrr committed:
Adding Custom Sentiment LRO task and its response

jessicawrr committed:
Adding Custom Sentiment task request and response example

kristapratico commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1651,6 +1659,62 @@
+    "CustomSentimentAnalysisTaskParameters": {
+      "type": "object",
+      "description": "Supported parameters for a Custom Sentiment Analysis task.",
+      "properties": {
```
Will CustomSentimentAnalysis support opinion mining (now or in the future)?

jessicawrr replied:
not in the near future. That's why we don't introduce the om parameter here

kristapratico commented on dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json:
```diff
@@ -0,0 +1,34 @@
+  "responses": {
+    "202": {
+      "headers": {
+        "Operation-Location": "{Endpoint}/language/analyze-text/jobs/{jobId}?api-version=2023-04-01"
```
Should the api version be 2023-04-15-preview?

```suggestion
        "Operation-Location": "{Endpoint}/language/analyze-text/jobs/{jobId}?api-version=2023-04-15-preview"
```

jessicawrr replied:
Good catch!! Will update

deyaaeldeen commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1651,6 +1659,62 @@
+          "items": {
+            "allOf": [
+              {
+                "$ref": "#/definitions/SentimentDocumentResult"
```
This object also describes the results for opinion mining and you said opinion mining is not part of this feature. This sharing could cause confusion. Could you please define your own result object that has things that are unique to this task only?

jessicawrr replied:
Sounds good. Will do that!

heaths replied:
@jessicawrr any update on resolving this issue?

deyaaeldeen replied:
@jessicawrr could you please address this comment?

jessicawrr replied:
Oh Sorry. I missed this. I've updated the PR and defined the custom sa specific result

heaths commented:
Please rebase your changes on the target release branch and don't merge main. It's impossible to review this:

![image](https://user-images.githubusercontent.com/1532486/222572731-3a0e5011-a2f3-4d2e-86cf-c24ea272d6f2.png)

```bash
git fetch origin release-cognitiveservices-Language-2023-04-15-preview
# on your topic branch...
git rebase origin/release-cognitiveservices-Language-2023-04-15-preview
git push --force
```

That will update this PR so please don't close this one and create a new one.

heaths commented:
Can you rebase this?

![image](https://user-images.githubusercontent.com/1532486/222594706-6164d238-027e-46e2-9efd-90d3738edfc8.png)

```bash
git fetch upstream release-cognitiveservices-Language-2023-04-15-preview
# on your topic branch
git rebase FETCH_HEAD
git push --force
```

markweitzel reviewed (approved):
## API Stewardship Board
Per the review on 2-Mar-23, this is good to go for preview. 
Note: Heath will partner with the team to clean up the branch.

heaths reviewed (changes requested):
Need to clean up the branch before merging. @Ransqure can you confirm that only the following two commits are needed?

```
commit df892ba48f2c2788d5ce0fd84488f85d572c918f
Author: Ranran Wang <rawan@microsoft.com>
Date:   Sun Feb 5 20:56:59 2023 -0800
    Adding Custom Sentiment task request and response example

commit f61a5f14a0a73cab833f9deba2c17a0dc8d1a629
Author: Ranran Wang <rawan@microsoft.com>
Date:   Sun Feb 5 15:52:57 2023 -0800
    Adding Custom Sentiment LRO task and its response
```

jessicawrr committed:
Fix the api version

heaths reviewed (changes requested):
@deyaaeldeen called out a change needed before I can merge.

heaths commented:
Please also fix the prettier issue. See https://github.com/Azure/azure-rest-api-specs/pull/22466/checks?check_run_id=11982254371 for details.

As for breaking changes ([policy](https://aka.ms/AzBreakingChangesPolicy)),

1. Don't add `type: object` to previously GA'd models. It would've been better to have them initially but, practically, it doesn't make a difference for our code generators. Adding them now is considered a breaking change.
2. Based on [the diff to 2022-05-01](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL3Jhdy5naXRodWJ1c2VyY29udGVudC5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvbWFpbi9zcGVjaWZpY2F0aW9uL2NvZ25pdGl2ZXNlcnZpY2VzL2RhdGEtcGxhbmUvTGFuZ3VhZ2Uvc3RhYmxlLzIwMjItMDUtMDEvYW5hbHl6ZXRleHQuanNvbiIsInRhcmdldFVybCI6Imh0dHBzOi8vcmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy8yYzMwODQxZGYzYWVjZWM2NWEzNjM0ZTMwYWEyMDVlOTNmOTkyZDkxL2Rldi9jb2duaXRpdmVzZXJ2aWNlcy9kYXRhLXBsYW5lL0xhbmd1YWdlL2FuYWx5emV0ZXh0Lmpzb24ifQ==), the newly added required properties on previously GA'd models are all on response models, so allowed by the Breaking Change Policy since this is a new api-version. That said, some models updated `$ref`s to point to similar models in _common.json_. at least from what I can see in the [corresponding _common.json_](https://github.com/Azure/azure-rest-api-specs/blob/2c30841df3aecec65a3634e30aa205e93f992d91/dev/cognitiveservices/data-plane/Language/common.json) these appear to be equivalent. Most of these validation errors may be from between the new and old versions of 2023-04-15-preview, which validation shouldn't really be running on anyway since it's **in-development**. I opened Azure/azure-sdk-tools#5713 to track also adding links or at least a model name to make it more obvious where the validation errors are.

deyaaeldeen commented:
What is the status for this PR?

heaths commented:
`npx prettier@2.6.2 -c dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json` shows no issues on the checked out files for this PR. I'll try to run failed checks again; otherwise, need @lmazuel or @weshaggard to use their special powers to merge. This is not a check I can override with a label.

jessicawrr commented:
Hi everyone. Sorry for the late response. We need to merge this PR in. I believe I address all the comments. @heaths is there anything else need to address before merging it?

heaths commented:
@jessicawrr you'll need to resolve the conflict in analyzetext.json, but @lmazuel or @weshaggard will have to merge. I can't because the PrettierCheck is required and I can't determine why. There's no output in the logs, and when I try it locally against this PR I also see no output. No idea what's wrong.

weshaggard commented:
@heaths what about the breaking change?

heaths commented:
> what about the breaking change?

@weshaggard they are all in response models where `required` doesn't make much difference in most languages' code generators, but where for newly-added properties anyway.

jessicawrr committed:
Define custom SA specific result object

jessicawrr committed:
Refer to custom SA specific result object

jessicawrr committed:
Merge branch 'release-cognitiveservices-Language-2023-04-15-preview' into user/rawan/customsa-runtime-api

jessicawrr commented:
@heaths and @deyaaeldeen I addressed the comments you point out. Also. I've resolved the conflict in analyzetext.json. I believe we are good to merge it :)

weshaggard commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -309,6 +315,7 @@
         "CustomSingleLabelClassification",
         "CustomMultiLabelClassification",
         "AbstractiveSummarization",
+        "CustomSentimentAnalysis"
```
```suggestion
        "CustomSentimentAnalysis",
```

weshaggard commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -349,6 +356,7 @@
         "CustomSingleLabelClassificationLROResults",
         "CustomMultiLabelClassificationLROResults",
         "AbstractiveSummarizationLROResults",
+        "CustomSentimentAnalysisLROResults"
```
```suggestion
        "CustomSentimentAnalysisLROResults",
```

weshaggard committed:
Add missing comma to fix json file

heaths commented:
There an unresolvable reference that must be fixed before we can merge or it'll break everything for this service directory.

deyaaeldeen commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1392,6 +1392,89 @@
+        "sentences"
+      ]
+    },
+    "CustomSentimentSentenceResult": {
```
nitpick: Would be nice if we align the custom names with the prebuilt names.

deyaaeldeen commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1392,6 +1392,89 @@
         "documents"
       ]
     },
+    "CustomSentimentResult": {
```
Is this an orphan definition? I don't see it referenced

jessicawrr committed:
Align the custom sa result name with prebuild ones

jessicawrr committed:
Refer the new object

mssfang commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -168,6 +168,9 @@
           "Successful Healthcare DocumentType Post Request": {
             "$ref": "./examples/text/SuccessfulHealthcareDocumentTypePostRequest.json"
+          },
+          "Successful Custom Custom Sentiment Analysis Post Request": {
```
NIT: "Custom" is repeating.

mssfang commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1645,6 +1736,62 @@
+      },
+      "allOf": [
+        {
+          "$ref": "#/definitions/CustomResult"
```
should we align with other same ref. 
See https://github.com/Azure/azure-rest-api-specs/blob/c34abf84ee5f0e20f599036184e9817da49d1945/dev/cognitiveservices/data-plane/Language/analyzetext.json#L4094


 - $ref: common.json#/definitions/CustomResult

mssfang commented on dev/cognitiveservices/data-plane/Language/analyzetext.json:
```diff
@@ -1645,6 +1736,62 @@
+      },
+      "allOf": [
+        {
+          "$ref": "#/definitions/CustomTaskParameters"
```
See https://github.com/Azure/azure-rest-api-specs/blob/c34abf84ee5f0e20f599036184e9817da49d1945/dev/cognitiveservices/data-plane/Language/analyzetext.json#L586

jessicawrr committed:
Resolve reference

jessicawrr committed:
Resolve reference

jessicawrr committed:
Resolve reference

konrad-jamrozik commented:
> `npx prettier@2.6.2 -c dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskSubmit.json` shows no issues on the checked out files for this PR. I'll try to run failed checks again; otherwise, need @lmazuel or @weshaggard to use their special powers to merge. This is not a check I can override with a label.

@heaths looks like the problem is with a different file: [`*Request.json` instead of `*Submit.json`](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2684744&view=logs&j=781fd7cb-cbf2-5d4a-a595-f7c2fa08ccdd&t=32535914-32d8-5ecd-c7d0-aa49d9c79250&l=15):

``` powershell
npx prettier -c "dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskStatusRequest.json"
Checking formatting...
[warn] dev\cognitiveservices\data-plane\Language\examples\text\SuccessfulCustomSentimentAnalysisTaskStatusRequest.json
[warn] Code style issues found in the above file. Forgot to run Prettier?
```

@jessicawrr you can fix the issue by running following command in the local clone of your fork on branch `user/rawan/customsa-runtime-api`, and pushing the resulting changes:

``` powershell
npx prettier -w "dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomSentimentAnalysisTaskStatusRequest.json"
```

jessicawrr commented:
I see. Let me try

heaths commented:
@konrad-jamrozik @weshaggard the netcoreapp3.1 issue is still a blocker. It can't run the breaking change validation until that's fixed or this PR is rebased, which it may need to be anyway. Let me know, and I can help @jessicawrr if needed.

@jessicawrr  you also need to resolve the conflicts since another version of _analyzetext.json_ went in already.

<img width="635" alt="image" src="https://user-images.githubusercontent.com/1532486/231598912-56aeb4a5-d1f1-4d29-a1b9-88ed1a810f27.png">

weshaggard commented:
/azp run

weshaggard commented:
@heaths that should be fixed when a new pipeline is triggered after the merge conflicts are resolved.

jessicawrr committed:
Merge branch 'release-cognitiveservices-Language-2023-04-15-preview' into user/rawan/customsa-runtime-api

jessicawrr committed:
fix pretty break

jessicawrr committed:
Merge branch 'user/rawan/customsa-runtime-api' of https://github.com/jessicawrr/azure-rest-api-specs into user/rawan/customsa-runtime-api

jessicawrr commented:
@heaths I believe I resolved everything blocking the merge (resolving the conflict and fix the PrettierCheck). Please take a look.
